hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsOutputStream.java

```diff
@@ -697,12 +848,76 @@ private synchronized void shrinkWriteOperationQueue() throws IOException {
     }
   }
 
-  private static class WriteOperation {
-    private final Future<Void> task;
-    private final long startOffset;
-    private final long length;
+  /**
+   * Flushes the written bytes to the Azure Blob Storage service.
+   *
+   * @param offset                the offset up to which data needs to be flushed.
+   * @param retainUncommittedData whether to retain uncommitted data after flush.
+   * @param isClose               whether this flush is happening as part of a close operation.
+   * @throws IOException if an I/O error occurs.
+   */
+  private synchronized void flushWrittenBytesToServiceInternal(final long offset,
+      final boolean retainUncommittedData, final boolean isClose)
+      throws IOException {
+
+    // Flush is called for append blob only on close
+    if (this.isAppendBlob && !isClose) {
+      return;
+    }
+
+    // Tracker to monitor performance metrics
+    AbfsPerfTracker tracker = client.getAbfsPerfTracker();
 
-    WriteOperation(final Future<Void> task, final long startOffset, final long length) {
+    // Performance information for tracking this method's performance
+    try (AbfsPerfInfo perfInfo = new AbfsPerfInfo(tracker,
+        "flushWrittenBytesToServiceInternal", "flush")) {
+
+      AbfsRestOperation op;
+
+      try {
+        // Attempt to flush data to the remote service
+        op = remoteFlush(offset, retainUncommittedData, isClose, leaseId,
+            tracingContext);
+      } catch (InvalidIngressServiceException ex) {
+        // If an invalid ingress service is encountered, switch handler and retry
+        switchHandler();
+        op = remoteFlush(offset, retainUncommittedData, isClose, leaseId,
+            tracingContext);
```

Comment:
So DFS today calls flush even if there are no appends, in case of Blob handling we have taken care of this by not doing any PutBlockList calls if map is empty. Since we are planning not to change the DFS behaviour, this handling is needed.